### PR TITLE
Subordinate charms handling.

### PR DESCRIPTION
**Pull request by esunar**
It is ready for initial review. 
All the subordinate charms that have openstack tracks are added.
Added a general purpose is_charm_supported function

### CODE REVIEW

File: cou/apps/subordinate.py
Comment (gabrielcocenza) [inline cou/apps/subordinate.py:55]: maybe something like this would be more clear:

```python
@property
def channel(self):
    return self._channel

@channel.setter
def channel(self, charm_channel):
    try:
         OpenStackRelease(charm_channel.split("/")[0])
         self._channel = charm_channel
    except ValueError as exc:
         ....
```

File: cou/apps/subordinate.py
Comment (gabrielcocenza) [inline cou/apps/subordinate.py:74]: IMO `ApplicationError` is more appropriate in that case. If an O7K subordinate don't have the channel as expected, I think we can stop the upgrading process

File: cou/utils/openstack.py
Comment (gabrielcocenza) [inline cou/utils/openstack.py:355]: maybe `is_charm_version_supported` represent better what we are trying to do because we support a group of charms, but some of then we don't support version, right?

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:74]: ok

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:55]: ok

File: cou/utils/openstack.py
Comment (esunar) [reply cou/utils/openstack.py:355]: ok

File: cou/apps/subordinate.py
Comment (gabrielcocenza) [inline cou/apps/subordinate.py:49]: Don't we need to also refresh the current channel?
It might be good also explaining that we are not updating packages because the principal unit will do it.

File: cou/apps/app.py
Comment (agileshaw) [inline cou/apps/app.py:83]: I don't think we need to check for `charm not in SUBORDINATES` here. If `is_charm_supported(charm)`, then we already know the charm is either principal or a known subordinate.

File: cou/apps/subordinate.py
Comment (agileshaw) [inline cou/apps/subordinate.py:37]: This docstring is not explanatory. We could change it to something like:
```suggestion
        """Infer the OS release from subordinate charm's channel.
        
        We cannot determine the OS release base on workload packages because the principal charm has already upgraded the packages.
        :return: OpenStackRelease object.
        :rtype: OpenStackRelease
        """
```

Also added the `:return` variable descirption in docstring

File: cou/apps/subordinate.py
Comment (agileshaw) [inline cou/apps/subordinate.py:31]: There are `channel` getter/setter methods created later in the class, why do we still need a `self.channel` variable here?

File: cou/apps/subordinate.py
Comment (agileshaw) [inline cou/apps/subordinate.py:80]: Since now we are adopting the CLI guideline, please add a message to the exception

**Review by agileshaw — CHANGES_REQUESTED** [review]

File: cou/apps/subordinate.py
Comment (gabrielcocenza) [reply cou/apps/subordinate.py:31]: We are validating `channel` and for that you can see that the setter accepts a value named `charm_channel`. That is why is still needed. This [article](https://www.programiz.com/python-programming/property) has an example of that.

File: cou/apps/app.py
Comment (esunar) [reply cou/apps/app.py:83]: hacluster charm is an example

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:80]: ok

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:31]: I need to revert this change because when creating object channel is set to "" and it creates exception since we do not allow this.

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:37]: done

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:49]: added refresh

File: cou/apps/subordinate.py
Comment (gabrielcocenza) [reply cou/apps/subordinate.py]: I think using getter/setter is more appropriate. Maybe if we remove `channel` as a required parameter to instantiate `OpenStackApplication` and do it on `__post_init__` is a better solution. I believe the same behavior that you were using on ` OpenStackSubordinateApplication` should be applied on `OpenStackApplication`. Changing on `OpenStackApplication` would fix for both.

File: cou/utils/openstack.py
Comment (gabrielcocenza) [inline cou/utils/openstack.py:90]: I think [neutron-api-plugin-arista](https://docs.openstack.org/charm-guide/latest/project/charm-delivery.html) can be also included here

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:31]: done

File: cou/utils/openstack.py
Comment (esunar) [reply cou/utils/openstack.py:91]: added

File: cou/apps/app.py
Comment (agileshaw) [reply cou/apps/app.py:83]: I was misled by the docstring change in `is_charm_version_supported` function. I though the lookup file only contains principal charms because of the word "primary". I suggestion changing `primary` to `core` in that function's docstring.

But still, I think this is a very confusing approach. `is_charm_supported == true` should indicate the charm is supported, so we should not need to have a condition afterwards to exclude some charms. I suggest either remove hacluster from the lookup file till we are ready to support it, or have a EXCLUDES list in openstack.py to exclude some charms and add `charm not in EXCLUDES` in `is_charm_supported` function.

File: cou/apps/subordinate.py
Comment (agileshaw) [inline cou/apps/subordinate.py:88]: According to the CLI guideline, `Cannot` is preferred in this case
```suggestion
                f"Cannot determine the OpenStack version from channel: {charm_channel}: {exc}"
```

File: cou/apps/app.py
Comment (esunar) [reply cou/apps/app.py:83]: I fixed the docstring. We can talk the details.

Comment (esunar) [reply cou/apps/app.py]: removed hacluster from csv file and no need for that block.

File: cou/apps/subordinate.py
Comment (aieri) [inline cou/apps/subordinate.py:54]: This message should be made consistent with the one from the superclass

File: cou/apps/subordinate.py
Comment (aieri) [inline cou/apps/subordinate.py:75]: when do we need to set the channel?

File: cou/steps/plan.py
Comment (aieri) [inline cou/steps/plan.py:48]:
```suggestion
    principal_upgrade_plan = await create_upgrade_group(
```

File: cou/steps/plan.py
Comment (aieri) [inline cou/steps/plan.py:54]:
```suggestion
    plan.add_step(principal_upgrade_plan)
```

File: cou/steps/plan.py
Comment (aieri) [inline cou/steps/plan.py:87]: you called it filter_function, you might as well pass it to a filter ;)
```suggestion
    for app in filter(filter_function, analysis_result.apps):
```

**Review by aieri — CHANGES_REQUESTED** [review]

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:29]: done

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:39]: ok

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:54]: ok

File: cou/steps/plan.py
Comment (esunar) [reply cou/steps/plan.py:48]: ok

File: cou/steps/plan.py
Comment (esunar) [reply cou/steps/plan.py:54]: ok

File: cou/steps/plan.py
Comment (esunar) [reply cou/steps/plan.py:87]: ok

File: cou/utils/openstack.py
Comment (esunar) [reply cou/utils/openstack.py:356]: ok

File: cou/utils/openstack.py
Comment (esunar) [reply cou/utils/openstack.py]: fixed

File: cou/utils/openstack.py
Comment (esunar) [reply cou/utils/openstack.py]: fixed

File: cou/utils/openstack.py
Comment (esunar) [reply cou/utils/openstack.py]: now it is consistent

**Review by aieri — CHANGES_REQUESTED** [review]
LGTM, we're only missing moving the setter/getter of channel to the superclass as we had agreed

**Review by agileshaw — COMMENTED** [review]
For the functional code part, I just have one small suggestion inline. Haven't had the time to check the test part, will look into it next week.

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:73]: We cannot move them to superclass yet. Since OpenStack charms those are not principals sill exits. We cannot move them to superclass after pr #42

File: cou/utils/openstack.py
Comment (esunar) [reply cou/utils/openstack.py:361]: done

File: tests/unit/apps/test_app.py
Comment (gabrielcocenza) [inline tests/unit/apps/test_app.py:447]: Shouldn't we remove this unit test instead of skipping it?

File: tests/unit/conftest.py
Comment (gabrielcocenza) [inline tests/unit/conftest.py:153]:
```suggestion
    # OpenStack subordinate application
```

File: tests/unit/steps/test_steps_plan.py
Comment (gabrielcocenza) [inline tests/unit/steps/test_steps_plan.py:45]: I think it would be good to include `keystone-ldap` on `apps` to also show `Subordinate(s) upgrade plan`

File: tests/unit/apps/test_subordinate.py
Comment (gabrielcocenza) [inline tests/unit/apps/test_subordinate.py:62]:
```suggestion
def test_channel_setter(status):
```

File: tests/unit/apps/test_subordinate.py
Comment (gabrielcocenza) [inline tests/unit/apps/test_subordinate.py:69]: maybe we could have two tests using `parametrize`. One showing that `<openstack>/release` doesn't raise `ApplicationError` and other that does.

**Review by gabrielcocenza — CHANGES_REQUESTED** [review]

File: tests/unit/apps/test_app.py
Comment (esunar) [reply tests/unit/apps/test_app.py:447]: I can remove it after pr #42 is merged.

File: cou/apps/subordinate.py
Comment (rgildein) [inline cou/apps/subordinate.py:86]: I think we should remove `exc` form this message. It will be present in trace logs, since we are using `raise ApplicationError(...) from exc`.
Although we are not providing trace logs right now, but reported an issue #43 for it.

**Review by rgildein — COMMENTED** [review]
overall I think it's ok

File: tests/unit/steps/test_steps_plan.py
Comment (esunar) [reply tests/unit/steps/test_steps_plan.py:45]: done

File: tests/unit/apps/test_subordinate.py
Comment (esunar) [reply tests/unit/apps/test_subordinate.py:62]: done

File: tests/unit/apps/test_subordinate.py
Comment (esunar) [reply tests/unit/apps/test_subordinate.py:69]: done

File: cou/apps/subordinate.py
Comment (esunar) [reply cou/apps/subordinate.py:86]: done

**Review by gabrielcocenza — DISMISSED** [review]
LGTM

File: cou/apps/subordinate.py
Comment (aieri) [reply cou/apps/subordinate.py:73]: Will be tracked in https://github.com/canonical/charmed-openstack-upgrader/issues/45

**Review by aieri — APPROVED** [review]

File: tests/unit/apps/test_app.py
Comment (agileshaw) [reply tests/unit/apps/test_app.py:447]: I don't understand why we need to wait for #42 to be merged before removing this test since subordinates-related changes should be handled in this PR.

Even if we decide to leave the test here for now, the skip message is really confusing; should say something like "subordinate check is removed" or "added supports for some subordinates" etc

File: tests/unit/apps/test_subordinate.py
Comment (agileshaw) [inline tests/unit/apps/test_subordinate.py:88]: It's better to separate the success and failure cases into two test functions. The current logic is a bit coupled.

**Review by agileshaw — CHANGES_REQUESTED** [review]
I have two comments inline. I think we are ready to merge the PR after addressing them.

File: tests/unit/apps/test_subordinate.py
Comment (esunar) [reply tests/unit/apps/test_subordinate.py]: done

File: tests/unit/apps/test_app.py
Comment (esunar) [reply tests/unit/apps/test_app.py]: removed

File: tests/unit/apps/test_subordinate.py
Comment (agileshaw) [inline tests/unit/apps/test_subordinate.py:65]: Parameterize on a single variable doesn't need the parenthesis. Could you please remove them? (and also for the `invalid` test function)
```
@pytest.mark.parametrize( "channel", ["ussuri/stable", "victoria/stable", ...])

Comment (esunar) [reply tests/unit/apps/test_subordinate.py:65]: done

**Review by agileshaw — APPROVED** [review]
LGTM, thanks. Let's wait for the CI to pass and then merge.

**Review by aieri — APPROVED** [review]

**Pull request merged**
